File: dfs_bfs_backtracking/cpp/b3184.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stack>

int R, C;
std::vector<std::string> fields;
bool visited[250][250]{ 0 };

std::vector<std::pair<int, int>> directions{ {1, 0}, {-1, 0}, {0, 1}, {0, -1} };

void dfs(const int& row, const int& col, int& sheeps, int& wolves) {
	std::stack<std::pair<int, int>> stacks;
	stacks.push({ row, col });
	visited[row][col] = true;
	while (stacks.size()) {
		std::pair<int, int> check = stacks.top();
		stacks.pop();

		int& temp_row = check.first;
		int& temp_col = check.second;

		if (fields[temp_row][temp_col] == 'v') { wolves += 1; }		
		if (fields[temp_row][temp_col] == 'o') { sheeps += 1; }

		for (std::pair<int, int>& direction : directions) {
			if (temp_row + direction.first >= 0 &&
				temp_row + direction.first < R &&
				temp_col + direction.second >= 0 &&
				temp_col + direction.second < C &&
				!visited[temp_row + direction.first][temp_col + direction.second] &&
				fields[temp_row + direction.first][temp_col + direction.second] != '#') {
				visited[temp_row + direction.first][temp_col + direction.second] = true;
				stacks.push({ temp_row + direction.first, temp_col + direction.second });
			}
		}
	}
}

int main() {
	std::ios_base::sync_with_stdio(false);
	std::cin.tie(NULL);

	std::cin >> R >> C;
	fields.resize(R);

	for (int i = 0; i < R; ++i) {
		std::cin >> fields[i];
	}
	
	int sheeps = 0;
	int wolves = 0;

	for (int row = 0; row < R; ++row) {
		for (int col = 0; col < C; ++col) {
			if (fields[row][col] == '.' || fields[row][col] == 'o' || fields[row][col] == 'v') {
				if (!visited[row][col]) {
					int temp_sheeps = 0;
					int temp_wolves = 0;
					dfs(row, col, temp_sheeps, temp_wolves);
					temp_sheeps = temp_sheeps > temp_wolves ? temp_sheeps : 0;
					temp_wolves = temp_sheeps > temp_wolves ? 0 : temp_wolves;
					sheeps += temp_sheeps;
					wolves += temp_wolves;
				}
			}
		}
	}

	std::cout << sheeps << " " << wolves << std::endl;
}

File: dfs_bfs_backtracking/python/b3184.py
from sys import stdin

R, C = map(int, input().split())

def dfs(row, col):
    global visited
    temp_sheep = temp_wolves = 0
    directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    stacks = [[row, col]]
    visited[row][col] = 1
    while stacks:
        check = stacks.pop()
        if fields[check[0]][check[1]] == 'v':
            temp_wolves += 1
        if fields[check[0]][check[1]] == 'o':
            temp_sheep += 1

        for direction in directions:
            temp_row = check[0] + direction[0]
            temp_col = check[1] + direction[1]
            if temp_row >= 0 and temp_row < R and temp_col >= 0 and temp_col < C and not visited[temp_row][temp_col] and fields[temp_row][temp_col] != "#":
                visited[temp_row][temp_col] = 1
                stacks.append([temp_row, temp_col])
    return temp_sheep, temp_wolves

fields = [stdin.readline() for _ in range(R)]
visited = [[0 for _ in range(C)] for __ in range(R)]

sheep = wolves = 0

for row in range(R):
    for col in range(C):
        if fields[row][col] == "." or fields[row][col] == "v" or fields[row][col] == "o":
            if not visited[row][col]:
                temp_sheep, temp_wolves = dfs(row, col)
                temp_sheep = temp_sheep if temp_sheep > temp_wolves else 0
                temp_wolves = 0 if temp_sheep > temp_wolves else temp_wolves
                sheep += temp_sheep
                wolves += temp_wolves

print("{} {}".format(sheep, wolves))